{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3edecb68-9229-46f5-abeb-801a247c196e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: duckdb in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (1.1.2)\n",
      "Collecting ipython-sql\n",
      "  Obtaining dependency information for ipython-sql from https://files.pythonhosted.org/packages/30/8f/9e50fa53ffc371483f9d1b90c1175b706d28a2e978e90a8894035af01905/ipython_sql-0.5.0-py3-none-any.whl.metadata\n",
      "  Downloading ipython_sql-0.5.0-py3-none-any.whl.metadata (17 kB)\n",
      "Requirement already satisfied: pandas in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (2.2.3)\n",
      "Collecting prettytable (from ipython-sql)\n",
      "  Obtaining dependency information for prettytable from https://files.pythonhosted.org/packages/73/19/4bb9530512432774fdd7cb7c020851d4decbb811d95f86fd4f6a870a6d3e/prettytable-3.12.0-py3-none-any.whl.metadata\n",
      "  Downloading prettytable-3.12.0-py3-none-any.whl.metadata (30 kB)\n",
      "Requirement already satisfied: ipython in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from ipython-sql) (8.28.0)\n",
      "Requirement already satisfied: sqlalchemy>=2.0 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from ipython-sql) (2.0.36)\n",
      "Collecting sqlparse (from ipython-sql)\n",
      "  Obtaining dependency information for sqlparse from https://files.pythonhosted.org/packages/7a/13/5f6654c9d915077fae255686ca6fa42095b62b7337e3e1aa9e82caa6f43a/sqlparse-0.5.2-py3-none-any.whl.metadata\n",
      "  Downloading sqlparse-0.5.2-py3-none-any.whl.metadata (3.9 kB)\n",
      "Requirement already satisfied: six in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from ipython-sql) (1.16.0)\n",
      "Collecting ipython-genutils (from ipython-sql)\n",
      "  Obtaining dependency information for ipython-genutils from https://files.pythonhosted.org/packages/fa/bc/9bd3b5c2b4774d5f33b2d544f1460be9df7df2fe42f352135381c347c69a/ipython_genutils-0.2.0-py2.py3-none-any.whl.metadata\n",
      "  Downloading ipython_genutils-0.2.0-py2.py3-none-any.whl.metadata (755 bytes)\n",
      "Requirement already satisfied: numpy>=1.23.2 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from pandas) (2.1.2)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from pandas) (2.9.0.post0)\n",
      "Requirement already satisfied: pytz>=2020.1 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from pandas) (2024.2)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from pandas) (2024.2)\n",
      "Requirement already satisfied: typing-extensions>=4.6.0 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from sqlalchemy>=2.0->ipython-sql) (4.12.2)\n",
      "Requirement already satisfied: greenlet!=0.4.17 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from sqlalchemy>=2.0->ipython-sql) (3.1.1)\n",
      "Requirement already satisfied: decorator in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from ipython->ipython-sql) (5.1.1)\n",
      "Requirement already satisfied: jedi>=0.16 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from ipython->ipython-sql) (0.19.1)\n",
      "Requirement already satisfied: matplotlib-inline in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from ipython->ipython-sql) (0.1.7)\n",
      "Requirement already satisfied: prompt-toolkit<3.1.0,>=3.0.41 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from ipython->ipython-sql) (3.0.48)\n",
      "Requirement already satisfied: pygments>=2.4.0 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from ipython->ipython-sql) (2.18.0)\n",
      "Requirement already satisfied: stack-data in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from ipython->ipython-sql) (0.6.3)\n",
      "Requirement already satisfied: traitlets>=5.13.0 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from ipython->ipython-sql) (5.14.3)\n",
      "Requirement already satisfied: pexpect>4.3 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from ipython->ipython-sql) (4.9.0)\n",
      "Requirement already satisfied: wcwidth in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from prettytable->ipython-sql) (0.2.13)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.3 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from jedi>=0.16->ipython->ipython-sql) (0.8.4)\n",
      "Requirement already satisfied: ptyprocess>=0.5 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from pexpect>4.3->ipython->ipython-sql) (0.7.0)\n",
      "Requirement already satisfied: executing>=1.2.0 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from stack-data->ipython->ipython-sql) (2.1.0)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from stack-data->ipython->ipython-sql) (2.4.1)\n",
      "Requirement already satisfied: pure-eval in /home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages (from stack-data->ipython->ipython-sql) (0.2.3)\n",
      "Downloading ipython_sql-0.5.0-py3-none-any.whl (20 kB)\n",
      "Downloading ipython_genutils-0.2.0-py2.py3-none-any.whl (26 kB)\n",
      "Downloading prettytable-3.12.0-py3-none-any.whl (31 kB)\n",
      "Downloading sqlparse-0.5.2-py3-none-any.whl (44 kB)\n",
      "\u001b[2K   \u001b[38;2;114;156;31m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m44.4/44.4 kB\u001b[0m \u001b[31m4.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
      "\u001b[?25hInstalling collected packages: ipython-genutils, sqlparse, prettytable, ipython-sql\n",
      "Successfully installed ipython-genutils-0.2.0 ipython-sql-0.5.0 prettytable-3.12.0 sqlparse-0.5.2\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.2.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m24.3.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install duckdb ipython-sql pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "f1dd2fbd-afe3-462c-8af2-c4306fb4c1fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import duckdb\n",
    "\n",
    "# Connect to the DuckDB database\n",
    "conn = duckdb.connect('../hmp_microbiome.duckdb')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "66fd45bb-a8d4-4d48-9257-c1fc96dc14b2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   organism_id          accession                        organism_name  \\\n",
      "0            1  NZ_ACGB01000060.1              Acidaminococcus sp. D21   \n",
      "1            2      NZ_GG770410.1  Achromobacter piechaudii ATCC 43553   \n",
      "2            3      NZ_GL635869.1       Achromobacter xylosoxidans C54   \n",
      "3            5      NZ_GG704511.1            Acinetobacter sp. RUH2624   \n",
      "4            4      NZ_GG665950.1         Acinetobacter sp. ATCC 27244   \n",
      "\n",
      "                                            taxonomy  genome_size  \\\n",
      "0  Bacteria;Firmicutes;Negativicutes;Selenomonada...         8820   \n",
      "1  Bacteria;Proteobacteria;Betaproteobacteria;Bur...         2687   \n",
      "2  Bacteria;Proteobacteria;Betaproteobacteria;Bur...         1515   \n",
      "3  Bacteria;Proteobacteria;Gammaproteobacteria;Ps...        98692   \n",
      "4  Bacteria;Proteobacteria;Gammaproteobacteria;Ps...       357640   \n",
      "\n",
      "          processing_date                                         source_url  \n",
      "0 2024-10-24 08:04:29.982  https://ftp.ncbi.nih.gov/genomes/HUMAN_MICROBI...  \n",
      "1 2024-10-24 08:04:30.316  https://ftp.ncbi.nih.gov/genomes/HUMAN_MICROBI...  \n",
      "2 2024-10-24 08:04:32.568  https://ftp.ncbi.nih.gov/genomes/HUMAN_MICROBI...  \n",
      "3 2024-10-24 08:04:33.865  https://ftp.ncbi.nih.gov/genomes/HUMAN_MICROBI...  \n",
      "4 2024-10-24 08:04:33.411  https://ftp.ncbi.nih.gov/genomes/HUMAN_MICROBI...  \n"
     ]
    }
   ],
   "source": [
    "# Execute a query and fetch results as a Pandas DataFrame\n",
    "df = conn.execute(\"SELECT * FROM organisms LIMIT 5\").fetchdf()\n",
    "print(df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "19913ded-701b-49f8-9237-d34a4fcba7d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext sql"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "42d09d9c-6d50-4220-aa6c-dca1bd5fd63f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages/sql/connection.py\", line 45, in __init__\n",
      "    engine = sqlalchemy.create_engine(\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"<string>\", line 2, in create_engine\n",
      "  File \"/home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages/sqlalchemy/util/deprecations.py\", line 281, in warned\n",
      "    return fn(*args, **kwargs)  # type: ignore[no-any-return]\n",
      "           ^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages/sqlalchemy/engine/create.py\", line 550, in create_engine\n",
      "    entrypoint = u._get_entrypoint()\n",
      "                 ^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages/sqlalchemy/engine/url.py\", line 758, in _get_entrypoint\n",
      "    cls = registry.load(name)\n",
      "          ^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py\", line 375, in load\n",
      "    raise exc.NoSuchModuleError(\n",
      "sqlalchemy.exc.NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:duckdb\n",
      "\n",
      "Connection info needed in SQLAlchemy format, example:\n",
      "               postgresql://username:password@hostname/dbname\n",
      "               or an existing connection: dict_keys([])\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages/sql/magic.py\", line 196, in execute\n",
      "    conn = sql.connection.Connection.set(\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages/sql/connection.py\", line 70, in set\n",
      "    cls.current = existing or Connection(descriptor, connect_args, creator)\n",
      "                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages/sql/connection.py\", line 45, in __init__\n",
      "    engine = sqlalchemy.create_engine(\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"<string>\", line 2, in create_engine\n",
      "  File \"/home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages/sqlalchemy/util/deprecations.py\", line 281, in warned\n",
      "    return fn(*args, **kwargs)  # type: ignore[no-any-return]\n",
      "           ^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages/sqlalchemy/engine/create.py\", line 550, in create_engine\n",
      "    entrypoint = u._get_entrypoint()\n",
      "                 ^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages/sqlalchemy/engine/url.py\", line 758, in _get_entrypoint\n",
      "    cls = registry.load(name)\n",
      "          ^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/home/shawn/projects/data-pipeline-microbiom-duckdb/venv/lib/python3.11/site-packages/sqlalchemy/util/langhelpers.py\", line 375, in load\n",
      "    raise exc.NoSuchModuleError(\n",
      "sqlalchemy.exc.NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:duckdb\n",
      "\n",
      "Connection info needed in SQLAlchemy format, example:\n",
      "               postgresql://username:password@hostname/dbname\n",
      "               or an existing connection: dict_keys([])\n"
     ]
    }
   ],
   "source": [
    "%sql duckdb:///:hmp_microbiome.duckdb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ede0a093-9e8a-40d3-8100-8524fccd51a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_qualifiers</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>genes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>organisms</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>processing_log</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>sequences</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              name\n",
       "0  gene_qualifiers\n",
       "1            genes\n",
       "2        organisms\n",
       "3   processing_log\n",
       "4        sequences"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "conn.execute(\"SHOW TABLES\").fetchdf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "c88c0e2d-6fb4-4d11-856d-87b88a9bc435",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                             organism_name  gene_count\n",
      "0          Dysgonomonas gadei ATCC BAA-286         628\n",
      "1                Eggerthella sp. 1_3_56FAA           4\n",
      "2       Enterobacter hormaechei ATCC 49162         407\n",
      "3             Enterococcus faecalis TX1302          12\n",
      "4             Enterococcus faecalis TX2137           2\n",
      "..                                     ...         ...\n",
      "885    Corynebacterium accolens ATCC 49726           4\n",
      "886               Dermacoccus sp. Ellin185          12\n",
      "887                    Desmospora sp. 8437           4\n",
      "888           Desulfovibrio sp. 6_1_46AFAA           4\n",
      "889  Corynebacterium genitalium ATCC 33030        4575\n",
      "\n",
      "[890 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "# Set the number of threads for parallel processing\n",
    "conn.execute(\"PRAGMA threads=4;\")\n",
    "\n",
    "# Query to count genes per organism using parallel execution\n",
    "query = \"\"\"\n",
    "SELECT organism_name, COUNT(gene_id) AS gene_count \n",
    "FROM organisms o \n",
    "JOIN genes g ON o.organism_id = g.organism_id \n",
    "GROUP BY organism_name;\n",
    "\"\"\"\n",
    "result = conn.execute(query).fetchdf()\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "b8b4250f-1c39-488f-834d-563411a1aff2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Query results exported to 'output.csv'\n"
     ]
    }
   ],
   "source": [
    "# Export query results to a CSV file\n",
    "query = \"\"\"\n",
    "COPY (SELECT * FROM organisms LIMIT 10) TO 'output.csv' WITH (FORMAT CSV, HEADER TRUE);\n",
    "\"\"\"\n",
    "conn.execute(query)\n",
    "print(\"Query results exported to 'output.csv'\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ec089089-6159-4ec6-82ec-171c4f80b809",
   "metadata": {},
   "outputs": [],
   "source": [
    "conn.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01196edf-5f2f-405b-a792-d75345125b55",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
